src/main/java/com/google/sps/DataGraph.java

```diff
@@ -171,70 +171,93 @@ public boolean mutateGraph(Mutation mut) {
     // Getting the corresponding graph nodes from the graph map
     GraphNode startNode = graphNodesMap.get(startName);
     GraphNode endNode = graphNodesMap.get(endName);
+    String error = "";
 
     switch (mut.getType()) {
       case ADD_NODE:
         {
-          // adding a duplicate node doesn't make any change
-          if (!graphNodesMap.containsKey(startName)) {
-            // New lone node is a root
-            roots.add(startName);
-            // Create a new node with the given name and add it to the graph and the map
-            GraphNode newGraphNode =
-                GraphNode.create(startName, new ArrayList<>(), Struct.newBuilder().build());
-            graph.addNode(newGraphNode);
-            graphNodesMap.put(startName, newGraphNode);
+          // Check whether node to be added is a duplicate
+          if (graphNodesMap.containsKey(startName)) {
+            error = "Add node: Adding a duplicate node " + startName + "\n";
```

Comment:
yea makes sense! For add edge and delete edge, is it okay to just define a local error variable since there could be 2 different parts to the message?